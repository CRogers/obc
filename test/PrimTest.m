MODULE PrimTest;

IMPORT Out;

VAR x: INTEGER;

PROCEDURE Foo(x: INTEGER) IS "PrimFoo";
(* CODE printf("foo: %d\n", arg[1].i); *)

PROCEDURE Baz(): INTEGER IS "PrimBaz";
(* CODE result.i = 42; *)

TYPE Handler = PROCEDURE (x: INTEGER): INTEGER;

PROCEDURE Bar(x: INTEGER): INTEGER;
BEGIN
  Out.String("bar: "); Out.Int(x, 0); Out.Ln;
  RETURN 66
END Bar;

PROCEDURE Bip(h: Handler): INTEGER IS "PrimBip";
(* CODE result = callback(arg[1].p, 1, 34); *)

PROCEDURE Dip(h: Handler): INTEGER IS "PrimDip";
(* CODE result = obcall(arg[1].p, cp, bp, sp, 1, 35); *)

BEGIN
  Foo(17);
  x := Bar(Baz());
  x := Bar(Bip(Bar));
  x := Bar(Dip(Bar))
END PrimTest.
