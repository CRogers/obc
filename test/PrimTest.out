foo: 17
bar: 42
bar: 34
bar: 66
bar: 35
bar: 66
